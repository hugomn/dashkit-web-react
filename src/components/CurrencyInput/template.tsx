import React from "react";
import { TextInput } from "grommet";
import { IProps, centsToCurrencyString, handleOnChange } from ".";

const CurrencyInput = ({ value, onChange }: IProps) => (
  <TextInput
    value={centsToCurrencyString(value)}
    onChange={handleOnChange(onChange)}
    placeholder={`${process.env.REACT_APP_CURRENCY}0,00`}
    data-testid="currency-input"
  />
);

export default React.memo(CurrencyInput);
