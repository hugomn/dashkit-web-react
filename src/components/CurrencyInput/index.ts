import template from "./template";

export interface IProps {
  value: string;
  onChange: (newValue: string) => void;
}

const DECIMAL_POINT_SEPARATOR = ",";

export const centsToCurrencyString = (cents?: string) => {
  if (!cents) return cents;
  const paddedCents = cents.padStart(3, "0");
  const decimalPlaceIndex = paddedCents.length - 2;
  return `${process.env.REACT_APP_CURRENCY}${paddedCents.substr(
    0,
    decimalPlaceIndex
  )}${DECIMAL_POINT_SEPARATOR}${paddedCents.substr(decimalPlaceIndex)}`;
};

export const currencyStringToCents = (currencyString: string) => {
  const digits = currencyString.replace(/\D/g, "");
  const number = parseInt(digits, 10);
  return Number.isNaN(number) ? "" : number.toString();
};

export const handleOnChange = (handler: IProps["onChange"]) => (
  changeEvent: React.ChangeEvent<HTMLInputElement>
) => handler(currencyStringToCents(changeEvent.target.value));

export default template;
