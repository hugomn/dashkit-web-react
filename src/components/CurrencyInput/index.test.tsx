import { cleanup, fireEvent, render } from "@testing-library/react";
import * as React from "react";
import CurrencyInput from ".";

describe("CurrencyInput component", () => {
  let originalCurrency: string | undefined;

  beforeAll(() => {
    originalCurrency = process.env.REACT_APP_CURRENCY;
    process.env.REACT_APP_CURRENCY = "🍕";
  });
  afterAll(() => {
    process.env.REACT_APP_CURRENCY = originalCurrency;
  });

  afterEach(cleanup);

  it("Should match snapshot", () => {
    const { container } = render(<CurrencyInput value="12345" onChange={jest.fn()} />);
    expect(container).toMatchSnapshot();
  });

  it("should trigger onChange with correctly parsed number", () => {
    const onChangeMock = jest.fn();
    const { getByTestId } = render(<CurrencyInput value=" " onChange={onChangeMock} />);
    fireEvent.change(getByTestId("currency-input"), { target: { value: "🍕1,00" } });
    expect(onChangeMock).toHaveBeenLastCalledWith("100");
    fireEvent.change(getByTestId("currency-input"), { target: { value: "🍕0,01" } });
    expect(onChangeMock).toHaveBeenLastCalledWith("1");
    fireEvent.change(getByTestId("currency-input"), { target: { value: "🍕17965,11" } });
    expect(onChangeMock).toHaveBeenLastCalledWith("1796511");
    fireEvent.change(getByTestId("currency-input"), { target: { value: "" } });
    expect(onChangeMock).toHaveBeenLastCalledWith("");
  });

  it("should correctly parse and render input number", () => {
    const { getByTestId, rerender } = render(<CurrencyInput value="100" onChange={jest.fn()} />);
    const myInputElem = getByTestId("currency-input") as HTMLInputElement;
    expect(myInputElem.value).toBe("🍕1,00");

    rerender(<CurrencyInput value="1796511" onChange={jest.fn()} />);
    expect(myInputElem.value).toBe("🍕17965,11");

    rerender(<CurrencyInput value="" onChange={jest.fn()} />);
    expect(myInputElem.value).toBe("");

    rerender(<CurrencyInput value="1" onChange={jest.fn()} />);
    expect(myInputElem.value).toBe("🍕0,01");
  });
});
