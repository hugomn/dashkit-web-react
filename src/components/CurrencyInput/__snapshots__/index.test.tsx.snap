// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CurrencyInput component Should match snapshot 1`] = `
<div>
  <div
    class="StyledTextInput__StyledTextInputContainer-sc-1x30a0s-1 kYQCpm"
  >
    <input
      autocomplete="off"
      class="StyledTextInput-sc-1x30a0s-0 kvYrcH"
      data-testid="currency-input"
      placeholder="🍕0,00"
      value="🍕123,45"
    />
  </div>
</div>
`;
